on-chain-nft/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+use on_chain_nft_io::*;
+use primitive_types::U256;
+
+#[derive(Debug, Default, NFTStateKeeper, NFTCore, NFTMetaState)]
+pub struct OnChainNFT {
+    #[NFTStateField]
+    pub token: NFTState,
+    pub token_id: TokenId,
+    pub owner: ActorId,
+    pub base_image: Vec<u8>,
+    pub layers: BTreeMap<LayerId, BTreeMap<ItemId, String>>,
+    pub nfts: BTreeMap<TokenId, BTreeMap<LayerId, ItemId>>,
+    pub nfts_existence: BTreeSet<String>,
+}
+
+static mut CONTRACT: Option<OnChainNFT> = None;
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
+    let config: InitOnChainNFT = msg::load().expect("Unable to decode InitOnChainNFT");
+
+    let nft = OnChainNFT {
+        token: NFTState {
+            name: config.name,
+            symbol: config.symbol,
+            base_uri: config.base_uri,
+            ..Default::default()
+        },
+        owner: msg::source(),
+        base_image: config.base_image.into_bytes(),
+        layers: config.layers,
+        ..Default::default()
+    };
+    CONTRACT = Some(nft);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn handle() {
+    let action: OnChainNFTAction = msg::load().expect("Could not load OnChainNFTAction");
+    let nft = CONTRACT.get_or_insert(Default::default());
+    match action {
+        OnChainNFTAction::Mint {
+            description,
+            token_metadata,
+        } => OnChainNFTCore::mint(nft, description, token_metadata),
+        OnChainNFTAction::Burn { token_id } => OnChainNFTCore::burn(nft, token_id),
+        OnChainNFTAction::Transfer { to, token_id } => NFTCore::transfer(nft, &to, token_id),
+        OnChainNFTAction::TransferPayout {
+            to,
+            token_id,
+            amount,
+        } => NFTCore::transfer_payout(nft, &to, token_id, amount),
+        OnChainNFTAction::Approve { to, token_id } => NFTCore::approve(nft, &to, token_id),
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn meta_state() -> *mut [i32; 2] {
+    let query: OnChainNFTQuery = msg::load().expect("failed to decode input argument");
+    let nft = CONTRACT.get_or_insert(OnChainNFT::default());
+    match query {
+        OnChainNFTQuery::TokenURI { token_id } => {
+            let encoded = OnChainNFTCore::token_uri(nft, token_id)
+                .expect("Error in reading OnChainNFT contract state");
+            gstd::util::to_leak_ptr(encoded)
+        }
+        OnChainNFTQuery::Base(query) => {
+            let encoded =
+                NFTMetaState::proc_state(nft, query).expect("Error in reading NFT contract state");
+            gstd::util::to_leak_ptr(encoded)
+        }
+    }
+}
+
+pub trait OnChainNFTCore: NFTCore {
+    fn mint(&mut self, description: BTreeMap<LayerId, ItemId>, metadata: TokenMetadata);
+    fn burn(&mut self, token_id: TokenId);
+    fn token_uri(&mut self, token_id: TokenId) -> Option<Vec<u8>>;
+}
+
+impl OnChainNFTCore for OnChainNFT {
+    fn mint(&mut self, description: BTreeMap<LayerId, ItemId>, metadata: TokenMetadata) {
```

Comment:
if the levels are already stored in a certain order, then maybe it just makes sense to give a variant of these levels in that order? 
```suggestion
    fn mint(&mut self, description: Vec<ItemId>, metadata: TokenMetadata) {
```

Reply:
It will also make it easier to check for identical tokens